### DISCUSSION TITLE
Test user accounts for email OTP

### DISCUSSION THREAD

**Discussion by WesleyYue** [supabase/supabase#10298] [Questions]
I have an app that is magic link login via email only, and I want to create a test login with a static OTP so that our tests can just use this login. There are a few questions about this for phone numbers (which didn't really have a Supabase-offered solution). The closest thing I found was this: https://github.com/supabase/supabase/discussions/5358

I tried implementing this with a trigger on email instead, and I added the confirmation_sent_at = now() per the comments, but this still gave me the "Token has expired or is invalid" error. Is there something special about the email OTP verification that makes this trigger hack not work? If so, what is the recommended way to test a system that only has email login links as a method?

**Comment by WesleyYue** [comment]
Figured it out.

`confirmation_token` is exclusively for SMS OTP, so the entry we want to actually overwrite is `recovery_token` and `recovery_sent_at`.

Steps to implement:
1. Attempt to login with your test email with a locally hosted supabase. ex: "fake@yourdomain.com"
2. Go to supabase studio (localhost:54323) to check what the hashed recovery_token is for fake@yourdomain.com. Copy this.
3. Go to the mail server to check what your generated OTP was (localhost:54324)
4. Create this trigger

```
CREATE FUNCTION test_credentials()
RETURNS TRIGGER
AS $$
BEGIN
    IF (NEW.email = ANY(array['fake@yourdomain.com', '<test-phone2>'])) THEN
        NEW.recovery_token := '<copied-from-step2>';
        NEW.recovery_sent_at := now();
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

CREATE TRIGGER test_credentials BEFORE INSERT OR UPDATE ON auth.users FOR EACH ROW EXECUTE PROCEDURE test_credentials();
```

5. Any login with fake@yourdomain.com will have its hashed recovery token overwritten to match the OTP generated in step 3. So you can use that OTP repeatedly to login.

**Reply by WesleyYue** [reply]
Actually, this doesn't quite work. For some reason, if you try doing any API operations after logging in the second time (ie. not creating an account with the bypass token), the supabase client doesn't send the proper JWT with any API calls to the DB, even though getSession and getUser shows that it looks like it is authenticated. I have to forcefully spawn another createClient instance with the access token as an option for it to work. Maybe the supabase team can chime in as to why that is.

**Reply by kangmingtay** [reply]
Hey @WesleyYue, it feels like there are multiple issues you are facing here which makes it slightly confusing for me to understand fully since I don't have the full context. 

Seems like there are 2 main problems you are facing:

1. >  I want to create a test login with a static OTP so that our tests can just use this login.

It seems like what you're doing is that you're trying to use the trigger to hardcode the value of the `recovery_token` and `recovery_sent_at`. I'm guessing you are using steps 1-3 to basically set up the "hardcoded value" of the hashed recovery token and the actual value so that on subsequent logins, as long as the email and OTP matches, the `recovery_token` will be overwritten with the hashed recovery token? 

2. > the supabase client doesn't send the proper JWT with any API calls to the DB, even though getSession and getUser shows that it looks like it is authenticated

Can you describe where you are creating the supabase client? Is it being created in a server-side environment? If possible, can you show some code snippets to explain the following:

* How are you initialising the Supabase client? What options do you pass into it?
* Are you creating the Supabase client in the browser or in a server-side environment?

**Reply by WesleyYue** [reply]
> Hey @WesleyYue, it feels like there are multiple issues you are facing here which makes it slightly confusing for me to understand fully since I don't have the full context.
> 
> Seems like there are 2 main problems you are facing:
> 
> 1. > I want to create a test login with a static OTP so that our tests can just use this login.
> 
> It seems like what you're doing is that you're trying to use the trigger to hardcode the value of the `recovery_token` and `recovery_sent_at`. I'm guessing you are using steps 1-3 to basically set up the "hardcoded value" of the hashed recovery token and the actual value so that on subsequent logins, as long as the email and OTP matches, the `recovery_token` will be overwritten with the hashed recovery token?

Correct. The original problem is: logging in for e2e tests for an application that only has email magic link as a auth method.

> 
> 2. > the supabase client doesn't send the proper JWT with any API calls to the DB, even though getSession and getUser shows that it looks like it is authenticated
> 
> Can you describe where you are creating the supabase client? Is it being created in a server-side environment? If possible, can you show some code snippets to explain the following:
> 
> * How are you initialising the Supabase client? What options do you pass into it?
> * Are you creating the Supabase client in the browser or in a server-side environment?

I'm creating the supabase client in a Playwright test fixture (so technically "server-side"). Something like this:
```
    await supabase.auth.signInWithOtp({
      email: this.testCredEmail,
      options: {
        shouldCreateUser: false,
      },
    });

    const { data: loginData, error: otpError } = await supabase.auth.verifyOtp({
      email: this.testCredEmail,
      token: this.otpToken,
      type: "magiclink",
    });

    if (otpError) {
      throw new Error(`error verifying otp ${otpError}`);
    }

    if (loginData === undefined || loginData.session === undefined) {
      throw new Error(`loginData or loginData.session is undefined. loginData: ${loginData}`);
    }

    assert(loginData.user?.role === "authenticated");

    // For some reason we have to manually inject the JWT into the client. verifyOtp is not logging
    // in the client automatically.
    const supabase2 = createClient<Database>(supabaseUrl, supabaseAnonKey, {
    global: { headers: { Authorization: `Bearer ${loginData?.session?.access_token!}` } },
  });
    this.authenticatedSupabase = supabase2;
    ```

**Reply by kangmingtay** [reply]
can you share how you're instantiating the `supabase` object? what options are you passing into it? since this is running on a "server-side" environment, the default options like `persistSession` wouldn't work because there isn't a storage option available on the server-side and you'll need to explicitly set it to `false`

**Reply by WesleyYue** [reply]
Just vanilla, no options. But okay, I'll try the persistSession: false

`const supabase = createClient<Database>(supabaseUrl, supabaseAnonKey);`

**Reply by kangmingtay** [reply]
For reference, this is the [full list of options](https://github.com/supabase/supabase-js/blob/master/src/lib/types.ts#L10) you can specify for `createClient`. If you're using the client server-side, i would recommend setting `autoRefreshToken` and `persistSession` to false. By setting `persistSession` to false, the session will be stored in-memory instead. Let me know if it works for you!

**Reply by saravmajestic** [reply]
This is really helpful. I just had to minor modification to fix the 60s otp rate limit error
```CREATE OR REPLACE FUNCTION public.preset_OTP()
RETURNS TRIGGER
AS $$
BEGIN
    IF (NEW.email = ANY(array['fake@yourdomain.com', '<test-phone2>'])) THEN
        NEW.recovery_token := encode(sha224(concat(NEW.email,'123456')::bytea), 'hex');
        NEW.recovery_sent_at := NOW() - INTERVAL '2 minutes';
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE TRIGGER preset_OTP BEFORE INSERT OR UPDATE ON auth.users FOR EACH ROW EXECUTE PROCEDURE public.preset_OTP();
```

**Reply by unlocomqx** [reply]
Just a small fix at the beginning `CREATE OR REPLACE FUNCTION preset_OTP()`
> This is really helpful. I just had to minor modification to fix the 60s otp rate limit error
> 
> ```
> CREATE OR REPLACE FUNCTION preset_OTP()
> RETURNS TRIGGER
> AS $$
> BEGIN
>     IF (NEW.email = ANY(array['fake@yourdomain.com', '<test-phone2>'])) THEN
>         NEW.recovery_token := encode(sha224(concat(NEW.email,'123456')::bytea), 'hex');
>         NEW.recovery_sent_at := NOW() - INTERVAL '2 minutes';
>     END IF;
>     RETURN NEW;
> END;
> $$ LANGUAGE plpgsql;
> 
> CREATE OR REPLACE TRIGGER preset_OTP BEFORE INSERT OR UPDATE ON auth.users FOR EACH ROW EXECUTE PROCEDURE public.preset_OTP();
> ```

**Reply by rphly** [reply]
Hi guys, this doesn't seem to be workin for me and I'm not sure why. The user is getting created, but when I check auth.user, the `recovery_token` and `recovery_sent_at` "" and null.

**Reply by rphly** [reply]
Okay, I figured it out. The above solution by @saravmajestic works, just make sure that the email supplied in the array is all lowercase. The email is normalised by the Supabase client, while SQL is case-sensitive, so if you have uppercase letters in the email (as in random generated strings) it will fail.

**Reply by TDAK1509** [reply]
> Just a small fix at the beginning `CREATE OR REPLACE FUNCTION preset_OTP()`
> 
> > This is really helpful. I just had to minor modification to fix the 60s otp rate limit error
> > ```
> > CREATE OR REPLACE FUNCTION preset_OTP()
> > RETURNS TRIGGER
> > AS $$
> > BEGIN
> >     IF (NEW.email = ANY(array['fake@yourdomain.com', '<test-phone2>'])) THEN
> >         NEW.recovery_token := encode(sha224(concat(NEW.email,'123456')::bytea), 'hex');
> >         NEW.recovery_sent_at := NOW() - INTERVAL '2 minutes';
> >     END IF;
> >     RETURN NEW;
> > END;
> > $$ LANGUAGE plpgsql;
> > 
> > CREATE OR REPLACE TRIGGER preset_OTP BEFORE INSERT OR UPDATE ON auth.users FOR EACH ROW EXECUTE PROCEDURE public.preset_OTP();
> > ```

Very helpful, it works for me.

**Comment by diogeneshamilton** [comment]
Worked great for me! Thanks for figuring this out.

**Comment by gorkabalm** [comment]
It would be very very helpful to have this integrated in the Dashboard the same way it is for SMS OTP.

**Comment by Mohammad-R-Rashid** [comment]
Was working but now is no longer working after a Postgres update?

**Comment by alexmercerind** [comment]
Life saver. Thanks a lot man.
